contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,1705 @@
+                // console2.log(_root, "Selected and moved to right");
+                // currentNodePath[parentIndex] = _root;
+                parentIndex++;
+                return
+                    _selectTopCluster(node.right, searchNumber - index1, selectedNodes, pathsToSelectedNodes, currentNodePath, parentIndex);
+            } else {
+                revert(ClusterLib.ERROR_OCCURED_DURING_TRAVERSING_SELECTED_NODE);
+            }
+        }
+        // if not selected then, check if it lies between the indexes
+        else {
+            // console2.log("_root is not selected", _root);
+            // console2.log("searchNumber", searchNumber);
+            // _printArray("selected nodes this _root", selectedNodes);
+            (uint256 index1, uint256 index2, uint256 index3) = ClusterLib._getIndexesWithWeights(leftWeight, node.balance, rightWeight);
+
+            // console2.log("leftWeight", leftWeight);
+            // console2.log("node.balance", node.balance);
+            // console2.log("rightWeight", rightWeight);
+            // console2.log("index1", index1);
+
+            currentNodePath = currentNodePath._addAddressToEncodedArray(_root);
+
+            if (searchNumber <= index1) {
+                // console2.log(_root, "Not select and moved to left");
+                // currentNodePath[parentIndex] = _root;
+                parentIndex++;
+                return _selectTopCluster(node.left, searchNumber, selectedNodes, pathsToSelectedNodes, currentNodePath, parentIndex);
+            } else if (searchNumber > index1 && searchNumber <= index2) {
+                // console2.log(_root, "Wow!, Selected");
+                // console2.log("case 2");
+                // currentNodePath[parentIndex] = _root;
+                // parentIndex++;
+                return (_root, currentNodePath);
+            } else if (searchNumber > index2 && searchNumber <= index3) {
+                // console2.log(_root, "Not select and moved to right");
+                // currentNodePath[parentIndex] = _root;
+                parentIndex++;
+                return
+                    _selectTopCluster(node.right, searchNumber - index2, selectedNodes, pathsToSelectedNodes, currentNodePath, parentIndex);
+            } else {
+                revert(ClusterLib.ERROR_OCCURED_DURING_TRAVERSING_NON_SELECTED_NODE);
+            }
+        }
+    }
+
+    /// @notice When a node is selected, the left and right weights have to be reduced in memory
+    /// @param node Node to reduce the weights
+    /// @param selectedNodes List of selected nodes
+    /// @param pathsToSelectedNodes Paths to the selected nodes
+    /// @return leftWeight reduced left weight of the node
+    /// @return rightWeight reduced right weight of the node
+    function _getModifiedWeightes(
+        Node memory node,
+        address[] memory selectedNodes,
+        bytes[] memory pathsToSelectedNodes
```

Comment:
Encoding of paths to selected  nodes can be based on index from root. So a 2D array which tells possible values of selected nodes at a certain index from root will be more efficient

Reply:
This is not relevant anymore with the updated logic